test/functional/specs/C2579.js

```diff
@@ -62,14 +62,21 @@ const getIdentityCookieValue = request => {
 
 const instance1Config = () => configureAlloyInstance(altConfig);
 
-const instance1Event = ClientFunction(() =>
-  window.alloy("event", { data: { key: "value" } })
-);
+const instance1Event = ClientFunction(() => {
+  return new Promise(resolve => {
+    return window.alloy("sendEvent", { data: { key: "value" } }).then(resolve);
+  });
```

Comment:
Can you help me understand why you had to wrap it in a new promise?

Reply:
@Aaronius  So I thought this is the only way we can prevent a race condition, because we are resolving the promise within the `ClientFunction` since we can return a promise into testcafe to await it there. No? Crazy?

Reply:
I think there's a misunderstanding. If you return a promise, you can `await` it within your test, whether the promise is the one returned from Alloy or you wrap the Alloy one in another promise like you've done here. TestCafe has some special logic that says if the return value is a promise, it will wait for it and then serialize the resolved value.

The race condition problems we were having previously was because we were trying to return an object containing a promise, so TestCafe was serializing the object itself, and the promise never made it through the serialization/deserialization process because a promise can't be serialized.

Reply:
Ah yes, that makes more sense, thank you!